### ISSUE TITLE
Emulators as git submodules

### ISSUE THREAD

**Issue by kattjevfel** [beaumanvienna/marley#32]
Could the emulators be added as submodules instead of manually rebasing against commits? This way it'd be easier to update and perhaps what I'm more interested in: being able to more easily set the submodule directories elsewhere to reduce disk space wastage (when having multiple projects using that repo).

**Comment by beaumanvienna** [comment]
The short answer is that all five cores were modified so that they can only be used in Marley.

The long answer is that Marley's design follows a bundle approach. By that I mean everything is supposed to be as tight as possible integrated. With this approach Marley tries to make a difference compared to what is prevalent in the emulator community. Submodules that need to be extra loaded, plugins that need to be found and dynamically linked, online updates that are required for basic functions, separate emulator windows that are launched externally (and may or may not end up on the right display / on top of the fullscreen launcher window), resource and assets folders with fonts and maps and databases that need to be located dynamically, so all that modular stuff, adds little value to the end user. But it tends to break or cause friction at least. That's why Marley is using as little modular designs as possible. It is of course linked against dynamic libraries, but that's really it. The source is bundled in one package, there are no essential online updates for shared libs, plugins, shaders, etc, Marley does not allow pop-up windows and it does not depend on any external resources (instead it is using an internal resource system for fonts, plugins, and assets where files and entire zip archives get compiled into the application and that either loads the files directly into memory or into Marley's config folder). Regarding the external resources the firnware files are an exception of course. I wished I could add those to the project, but that would of course violate a lot of rules.
Regarding the technical aspect of using upstream versions for the emulator cores, all five were modified when I integrated them. First off, they are now static libraries. I had to rework the variable initialization for all of them to make it possible to call them as functions multiple times during the lifetime of the application. Basically, they were relying on memory that gets initialized when the OS loads the program, but not during program execution. Then the core modules render into Marley's main window instead of their own. So I changed or modified the OpenGL context for all cores. This also applies to the game controllers that are instantiated in the front end. Then I changed the mapping of hotkeys, so that the user can always use F5/F7 for saving/loading, f to toggle fullscreen, and ESC or the guide button to cancel a screen or operation. Last but not least, the cores sometimes share the same source code, so I had to rename namespaces and variables. Especially ppsspp and dolphin share a lot of the same code. But PCSX2 and Mupen64plus have also a dissambler for an RISC processor in common. 
For the rebase I diff the core module from when it was first committed to the current version and re-apply the same changes. It is a bit of a work which takes about an evening. 
So this was the long story. It is definitely one of the key questions in this project. Nicely observed!
